proposals/0522-or-patterns.rst

```diff
@@ -0,0 +1,427 @@
+programmers to explicitly match a list of constructors in a concise way. For the above
+function we get
+
+::
+
+    stringOfT :: T -> Maybe String
+    stringOfT (T1 s)        = Just s
+    stringOfT (T2{} ; T3{}) = Nothing
+
+This function doesn't match ``T4``, so we get our warning in the very first compile
+cycle or (even faster) in our IDE powered by a language server implementation.
+
+
+Proposed Change Specification
+-----------------------------
+
+Changes in the grammar
+~~~~~~~~~~~~~~~~~~~~~~
+
+We consider this as an extension to `Haskell 2010 grammar
+<https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-18000010.5>`_.
+The relevant non-terminal is ``apat``: ::
+
+  apat    →    var [ @ apat]                     (as pattern)
+          |    gcon                              (arity gcon  =  0)
+          |    qcon { fpat1 , … , fpatk }        (labeled pattern, k ≥ 0)
+          |    literal
+          |    _                                 (wildcard)
+          |    ( pat )                           (parenthesized pattern)
+          |    ( pat1 , … , patk )               (tuple pattern, k ≥ 2)
+          |    [ pat1 , … , patk ]               (list pattern, k ≥ 1)
+          |    ~ apat
+
+Or patterns extension adds one more production: ::
+
+          |    ( pat1 ; … ; patk )                (Or pattern, k ≥ 2)
+
+The ``;`` between the parentheses have (shift) priority that is lower than any other ``apat``'s (reduction) priority.
+
+Some examples that this new grammar produces: ::
+
+  -- in expression context
+  case e of
+    (T1 ; T2{} ; T3 a b) -> ...
+
+  -- in expression context
+  let ([x] ; (x : y : z)) = e1 in e2
+
+  -- pattern guards in declarations
+  f x y
+    | x@(T1 _ ; T2 a b) <- e1
+    , guard x
+    = e2
+
+  -- nested Or patterns
+  case e1 of
+    (((T1 ; T2) ; T3) ; T4) -> e2
+
+The new production doesn't add any ambiguities because of the mandatory parentheses, just like for tuples.
+
+NB: The new grammar allows Or patterns which bind variables. These will however be rejected in `2.2`_.
+
+.. _2.2:
+
+Static semantics of Or pattern matching
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Or patterns which bind variables are rejected in the renamer.
+
+
+We give the static semantics in terms of *pattern types*.
+
+A pattern type, as defined `here <https://mpickering.github.io/pattern-synonyms-extended.pdf>`__ (chapter 6), has the form ``req => prov => arg_1 -> ... -> arg_k -> res``, where ``req`` denotes required constraints and ``prov`` denotes provided constraints.
+
+
+When we have two patterns ``p1`` and ``p2`` with pattern types
+::
+
+    p1 :: forall xs1. Theta1 => prov1 => arg1_1 -> ... -> arg1_k -> res1,
+    p2 :: forall xs2. Theta2 => prov2 => arg2_1 -> ... -> arg2_l -> res2
+
+then the Or pattern ``(p1; p2)`` has the pattern type ::
+
+    (p1; p2) :: forall xs. (T1[xs/xs1], T2[xs/xs2]) => () => arg1_1[xs/xs1] -> ... -> arg_k[xs/xs1] -> res1[xs/xs1]
+
```

Comment:
I’m surprised you list `arg…` there. Unless I am mistaken, this would allow the following:
```
fromEither :: Either a a -> a
fromEither ((Left;Right) x) = x
```

Interesting, but probably not really what you want. Or am I missing something?

Reply:
That is indeed a good example and we discussed it today.

Originally, we left the definition as is because we wanted to allow arrows in `res`, which can happen with a view pattern like `((\f -> f 42) -> True) :: Int -> Bool`. Note that `Int -> Bool` is the `res` here.

So there is some notion of *arity* of a pattern type; Fortunately, the only patterns that will type-check according to the referenced paper are saturated ones. Note the rule for constructors and pattern synonyms in particular; For constructors, `res` must be an application of the DataCon's TyCon. For pattern synonyms, the arity is taken from the PS's definition. You may not undersaturate or oversaturate a PS app in a pattern. Saturation of view patterns and Or patterns follows from saturation at the leaves (DataCons and PatSyns).

In our changed semantics, we'd reject your example because `Left` and `Right` are not well-typed patterns to begin with.

To bring home the point, here are two different pattern synonyms with the same pattern type but different arities:

```hs
pattern P :: Int -> Bool
pattern P <- ((\f -> f 42) -> True)

pattern Q :: Int -> Bool
pattern Q n <- ((\b -> if b then 1 else 0) -> n)


f :: Bool -> ()
f (Q n) = ()

g :: Bool -> ()
g (P n) = () -- rejected! P has arity 0
```

Reply:
Ah that makes sense! Thanks for bearing me commenting without reading the paper carefully :-)